src/Service.GraphQLBuilder/GraphQLUtils.cs

```diff
@@ -270,5 +272,63 @@ public static ObjectType UnderlyingGraphQLEntityType(IType type)
+        /// <returns></returns>
+        public static string GetDataSourceNameFromGraphQLContext(IMiddlewareContext context, RuntimeConfig runtimeConfig)
+        {
+            string rootNode = context.Selection.Field.Coordinate.TypeName.Value;
+            string dataSourceName;
+
+            if (string.Equals(rootNode, "mutation", StringComparison.OrdinalIgnoreCase) || string.Equals(rootNode, "query", StringComparison.OrdinalIgnoreCase))
+            {
+                // we are at the root query node - need to determine return type and store on context.
+                // Output type below would be the graphql object return type - Books,BooksConnectionObject.
+                IOutputType outputType = context.Selection.Field.Type;
+                string entityName = outputType.TypeName();
+
+                // Below is only needed if say we have an Array return type or items object for plural case.
+                ObjectType underlyingFieldType = GraphQLUtils.UnderlyingGraphQLEntityType(outputType);
+
+                // Example: CustomersConnectionObject - for get all scenarios.
+                if (QueryBuilder.IsPaginationType(underlyingFieldType))
+                {
+                    IObjectField subField = GraphQLUtils.UnderlyingGraphQLEntityType(context.Selection.Field.Type).Fields[QueryBuilder.PAGINATION_FIELD_NAME];
+                    outputType = subField.Type;
+                    underlyingFieldType = GraphQLUtils.UnderlyingGraphQLEntityType(outputType);
+                    entityName = underlyingFieldType.Name;
+                }
+
+                // if name on schema is different from name in config.
+                // Due to possibility of rename functionality, entityName on runtimeConfig could be different from exposed schema name.
+                if (GraphQLUtils.TryExtractGraphQLFieldModelName(underlyingFieldType.Directives, out string? modelName))
+                {
+                    entityName = modelName;
+                }
+
+                dataSourceName = runtimeConfig.GetDataSourceNameFromEntityName(entityName);
+
+                // Store dataSourceName on context for later use.
+                context.ContextData.TryAdd(GenerateDataSourceNameKeyFromPath(context), dataSourceName);
+            }
+            else
+            {
+                // Derive node from path - e.g. /books/{id} - node would be books.
+                // for this queryNode path we have stored the datasourceName needed to retrieve query and mutation engine of inner objects
+                object? obj = context.ContextData[GenerateDataSourceNameKeyFromPath(context)];
+                dataSourceName = obj?.ToString()!;
+            }
+
+            return dataSourceName;
+        }
+
+        private static string GenerateDataSourceNameKeyFromPath(IMiddlewareContext context)
+        {
+            // prefix added to ensure uniqueness.
+            return $"Node_{context.Path.ToList().First()}";
```

Comment:
can you guarantee that context.Path has elements? if it has 0 elements `.First()` will throw an exception.
Also, perhaps just use list accessor `[0]` when Count > 0. There's a lot of extra work (casts) involved in `.First()` See https://github.com/dotnet/runtime/blob/72ff23bc9f631df174b9a60df82fe19c4d5ce1fa/src/libraries/System.Linq/src/System/Linq/First.cs#L65-L95

Reply:
Sure will just use the 0th element. 
<img width="451" alt="image" src="https://github.com/Azure/data-api-builder/assets/124841904/3f82c7f1-7a18-4795-8b40-88f31f590e44">
Path is always atleast populated with the root query node field that was being called if you see above.